doc/source/developing/releasing.rst

```diff
@@ -1,37 +1,41 @@
 How to Do a Release
 -------------------
 
-Periodically, the yt development community issues new releases. Since yt follows
-`semantic versioning <https://semver.org/>`_, the type of release can be read off
+Periodically, the yt development community issues new releases. yt loosely follows
+`semantic versioning <https://semver.org/>`_. The type of release can be read off
 from the version number used. Version numbers should follow the scheme
 ``MAJOR.MINOR.PATCH``. There are three kinds of possible releases:
 
 * Bugfix releases
 
-  These releases are regularly scheduled and will optimally happen approximately
-  once a month. These releases should contain only fixes for bugs discovered in
+  These releases should contain only fixes for bugs discovered in
   earlier releases and should not contain new features or API changes. Bugfix
-  releases should increment the ``PATCH`` version number. Bugfix releases should
+  releases only increment the ``PATCH`` version number. Bugfix releases should
   *not* be generated by merging from the ``main`` branch, instead bugfix pull
-  requests should be manually backported. Version ``3.2.2`` is a bugfix release.
+  requests should be backported to a dedicated branch.
+  See :ref:`doing-a-bugfix-release`. Version ``3.2.2`` is a bugfix release.
 
 * Minor releases
 
   These releases happen when new features are deemed ready to be merged into the
   ``stable`` branch and should not happen on a regular schedule. Minor releases
   can also include fixes for bugs if the fix is determined to be too invasive
-  for a bugfix release. Minor releases should *not* include
-  backwards-incompatible changes and should not change APIs.  If an API change
-  is deemed to be necessary, the old API should continue to function but might
-  trigger deprecation warnings. Minor releases should happen by merging the
-  ``main`` branch into the ``stable`` branch. Minor releases should increment the
-  ``MINOR`` version number and reset the ``PATCH`` version number to zero.
+  for a bugfix release. If an API change is deemed to be necessary, the old API
+  should continue to function but might trigger deprecation warnings. Minor
+  releases may also remove some previously deprecated functionalities on the
+  explicit condition that a warning was emitted for that piece of code in at
+  latest one previous minor release. Such warnings are required to be visible to
+  users by default (as opposed to builtin ``DeprecationWarning``), and state
+  when deprecation started, as well as point to the new prefered alternative,
+  see :ref:`how-to-deprecate`. Minor releases should happen by merging the
+  ``main`` branch into the ``stable`` branch. Minor releases should increment
+  the ``MINOR`` version number and reset the ``PATCH`` version number to zero.
   Version ``3.3.0`` is a minor release.
```

Comment:
Previous to yt 4.0 we didn't have a stated policy for actually *removing* deprecated stuff, so it was never actually removed (AFAICT), so yes, it's new in the sense that it aligns with what we're doing now, but it wasn't explicitly documented yet.

Reply:
I think it would be worthwhile to reach a consensus on the timescale over which we remove deprecated features, rather than just creating de facto policy in a docs PR.

Reply:
I mean I opened this discussion 2 years ago https://github.com/yt-project/yt/issues/2911
I seem to remember it was discussed in a triage meeting and we _did_ reach a consensus on the policy which is already expressed in the `issue_deprecation_warning` function. Here I'm just documenting it.

Reply:
To add to this, I find that promising an exact version for when we want to remove stuff in the future is counter productive: it's the reason why we haven't been able to cut a 4.1 version yet, almost a year after 4.0. Namely, we promised to make ambiguous field keys raise errors, but we're not there yet (though I'm working on it). A much safer option is to tell _that_ we plan to remove stuff, and try to keep it around for a couple minor versions (or longer if it's not too much of a burden), but in general I think we shouldn't try to predict the future regarding deprecations, but each deprecation cycle can be discussed in its own time.

Reply:
> I think it would be worthwhile to reach a consensus on the timescale over which we remove deprecated features, rather than just creating de facto policy in a docs PR.

Update on this: since #3603 (yt 4.1), no deprecation warnings has had a clear end date. This is intentional: that way we can live with deprecated code as long as it doesn't get in the way of major refactors or features (and we also don't break any promises by dropping it as soon as it does).

> Minor releases may also remove some previously deprecated functionalities on the
  explicit condition that a warning was emitted for that piece of code in at
  latest one previous minor release

To be clear, here I my intention was not to advocate for ever shorter deprecation cycles; I think that the longer they are, the better. I'm happy to rephrase this sentence in a way that better reflects that instead of focussing on the most agressive juriceprudence. How about this ?

"Minor releases may occasionally expire small deprecations, but expiration should only be a last resort if and when deprecated code becomes a maintenance burden."